### PR TITLE
Update cover crypt 12

### PR DESCRIPTION
**Pull request by Manuthor**
- [x] update crypto_core to 8.0
- [x] update cover_crypt 12.0
- [x] update ECIES to Ecies salsa seal box
- [x] replace aesgcm impl' with Aes256Gcm crypto_core impl' (with authenticated_data as parameter)

### CODE REVIEW

File: crates/aesgcm/src/core/aesgcm.rs [hunk 1]
```diff
@@ -1,78 +1,119 @@
...
+/// error occurred during encryption).
+pub fn encrypt(
+    key: [u8; Aes256GcmRust::KEY_LENGTH],
+    nonce: [u8; Aes256GcmRust::NONCE_LENGTH],
+    plaintext: &[u8],
+    authenticated_data: &[u8],
+) -> Result<Vec<u8>, AesGcmError> {
+    let key = SymmetricKey::try_from_bytes(key)?;
+    let nonce = Nonce::try_from_bytes(nonce)?;
+    Ok(Aes256GcmRust::new(&key).encrypt(&nonce, plaintext, Some(authenticated_data))?)
```
Comment (tbrezot) [inline crates/aesgcm/src/core/aesgcm.rs:47]: Why don't you manage type conversion in the interfaces? I am not sure this wrapper is of much use.

File: crates/aesgcm/src/core/aesgcm.rs [hunk 2]
```diff
@@ -1,78 +1,119 @@
...
-
-        // Transformed from a byte array:
-        let nonce = *Nonce::from_slice(nonce);
-
-        Ok(Self { cipher, nonce })
-    }
-
-    pub fn encrypt(&self, plaintext: &[u8]) -> Result<Vec<u8>, AesGcmError> {
-        Ok(self.cipher.encrypt(&self.nonce, plaintext)?)
+/// The `decrypt` function takes a key, nonce, ciphertext, and authenticated
```
Comment (tbrezot) [inline crates/aesgcm/src/core/aesgcm.rs:50]: Same: please shorten this doc. No one will ever read a 30-lines long documentation. Even more if this is to say "This function decrypts the given AES-256 GCM ciphertext using the provided key, nonce and authentication data".

File: crates/aesgcm/src/error.rs [hunk 3]
```diff
@@ -9,25 +10,22 @@ use wasm_bindgen::JsValue;
 pub enum AesGcmError {
     AesGcm(String),
     AesGcmInvalidLength(String),
```
Comment (tbrezot) [inline crates/aesgcm/src/error.rs:12]: Are you still using these members?

File: crates/aesgcm/src/ffi/aesgcm.rs [hunk 4]
```diff
@@ -13,28 +14,42 @@ unsafe extern "C" fn aesgcm(
...
+    let authenticated_data = ffi_read_bytes!(
+        "authenticated_data",
+        authenticated_data_ptr,
+        authenticated_data_len
     );
     let output = if encrypt_flag {
-        ffi_unwrap!(aesgcm.encrypt(input_data_bytes), "AESGCM encrypt error")
+        ffi_unwrap!(
+            encrypt(key, nonce, input_data_bytes, authenticated_data),
+            "AESGCM encrypt error"
```
Comment (tbrezot) [inline crates/aesgcm/src/ffi/aesgcm.rs:46]:
```suggestion
            "AES-256 GCM encryption error"
```

File: crates/aesgcm/src/ffi/aesgcm.rs [hunk 5]
```diff
@@ -13,28 +14,42 @@ unsafe extern "C" fn aesgcm(
...
+        )
     } else {
-        ffi_unwrap!(aesgcm.decrypt(input_data_bytes), "AESGCM decrypt error")
+        ffi_unwrap!(
+            decrypt(key, nonce, input_data_bytes, authenticated_data),
+            "AESGCM decrypt error"
```
Comment (tbrezot) [inline crates/aesgcm/src/ffi/aesgcm.rs:51]:
```suggestion
            "AES-256 GCM decryption error"
```

File: crates/aesgcm/src/lib.rs [hunk 6]
```diff
@@ -10,4 +10,4 @@ pub mod wasm_bindgen;
 mod core;
 mod error;
 
-pub use crate::core::aesgcm::{ReExposedAesGcm, BLOCK_LENGTH, KEY_LENGTH, NONCE_LENGTH};
+pub use crate::core::aesgcm::{decrypt, encrypt, BLOCK_LENGTH};
```
Comment (tbrezot) [inline crates/aesgcm/src/lib.rs:13]: Why do you need a `BLOC_LENGTH`? It seems to me that this is of no use for a user.

File: crates/cover_crypt/Cargo.toml [hunk 7]
```diff
@@ -17,8 +18,8 @@ wasm_bindgen = ["js-sys", "wasm-bindgen"]
 default = []
 
 [dependencies]
-cosmian_cover_crypt = { version = "11.0.1", features = ["serialization"] }
-cosmian_crypto_core = "7.0"
+cosmian_cover_crypt = { version = "12.0.0", features = ["serialization", "test_utils"] }
```
Comment (tbrezot) [inline crates/cover_crypt/Cargo.toml:21]: If you need the feature `test_utils`, import it as a dev-dependency.
```suggestion
cosmian_cover_crypt = { version = "12.0.0", features = ["serialization"] }
```

File: crates/cover_crypt/src/wasm_bindgen/tests.rs [hunk 8]
```diff
@@ -44,10 +42,8 @@ fn encrypt_header(
         ),
         "Error encrypting header"
     );
-    EncryptedHeader::try_from_bytes(
-        &encrypted_header.to_vec()[CoverCryptX25519Aes256::SYM_KEY_LENGTH..],
-    )
-    .map_err(|e| JsValue::from_str(&e.to_string()))
+    EncryptedHeader::deserialize(&encrypted_header.to_vec()[Aes256Gcm::KEY_LENGTH..])
```
Comment (tbrezot) [inline crates/cover_crypt/src/wasm_bindgen/tests.rs:45]: The size of the key should depend on Covercrypt. It seems this size is not exposed. Please open an issue to track this.

File: crates/ecies/src/ffi/tests.rs [hunk 9]
```diff
@@ -37,7 +42,8 @@ fn encrypt_decrypt() {
         );
 
         // FFI encrypt output
-        let mut ciphertext_bytes = vec![0u8; plaintext.len() + h_ecies_encrypt_get_overhead_size()];
+        let mut ciphertext_bytes =
+            vec![0u8; plaintext.len() + h_ecies_salsa_seal_box_get_encryption_overhead()];
```
Comment (tbrezot) [inline crates/ecies/src/ffi/tests.rs:46]: Can't you use:
```suggestion
            vec![0u8; plaintext.len() + EciesSalsaSealBox::ENCRYPTION_OVERHEAD];
```

**Review by tbrezot — COMMENTED** [review]
Great job, thanks!

File: Cargo.toml [hunk 10]
```diff
@@ -9,10 +9,12 @@ members = [
   "crates/findex",
   "crates/fpe",
 ]
+resolver = "2"
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply Cargo.toml]: done in https://github.com/Cosmian/cloudproof_rust/pull/34/commits/023dad20bc0044d1db6d8a22d64b9d991f46995b

File: crates/aesgcm/src/core/aesgcm.rs [hunk 11]
```diff
@@ -1,78 +1,119 @@
...
-
 pub const BLOCK_LENGTH: usize = 16;
 
-/// The `ReExposedAesGcm` struct contains an instance of the `Aes256Gcm` cipher
-/// and a nonce of length 12. The crate `aesgcm` has received one security audit
-/// by NCC Group, with no significant findings. We would like to thank
-/// `MobileCoin` for funding the audit.
+/// The `encrypt` function takes a key, nonce, plaintext, and authenticated data
+/// as input and returns the encrypted ciphertext using AES-256 GCM encryption
+/// in Rust.
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply crates/aesgcm/src/core/aesgcm.rs]: it has been simplified

File: crates/aesgcm/src/core/aesgcm.rs (hunk 1 above)
Comment (Manuthor) [reply crates/aesgcm/src/core/aesgcm.rs]: good point. Done here https://github.com/Cosmian/cloudproof_rust/pull/34/commits/83d5042e0a5c35c8c6962b6e310c1a42fe964220

File: crates/aesgcm/src/core/aesgcm.rs (hunk 2 above)
Comment (Manuthor) [reply crates/aesgcm/src/core/aesgcm.rs]: done

File: crates/aesgcm/src/error.rs (hunk 3 above)
Comment (Manuthor) [reply crates/aesgcm/src/error.rs]: removed

File: crates/aesgcm/src/ffi/aesgcm.rs (hunk 4 above)
Comment (Manuthor) [reply crates/aesgcm/src/ffi/aesgcm.rs]: done

File: crates/aesgcm/src/lib.rs (hunk 6 above)
Comment (Manuthor) [reply crates/aesgcm/src/lib.rs]: indeed, thanks

File: crates/cover_crypt/Cargo.toml (hunk 7 above)
Comment (Manuthor) [reply crates/cover_crypt/Cargo.toml]: indeed

File: crates/cover_crypt/src/wasm_bindgen/tests.rs (hunk 8 above)
Comment (Manuthor) [reply crates/cover_crypt/src/wasm_bindgen/tests.rs]: actually SYM_KEY_LENGTH is exposed

**Pull request merged**